File: Extras/Gaia.Ajax.GPL/4.0 - 31.10.2013/Source/Gaia.WebWidgets.CommunityExtensions/Widgets/FileUpload.cs
/*******************************************************************
 * Gaia Ajax - Ajax Control Library for ASP.NET  
 * Copyright (C) 2008 - 2011 Gaiaware AS
 * All rights reserved. 
 * This program is distributed under either GPL version 3 
 * as published by the Free Software Foundation or the
 * Gaia Commercial License version 1 as published by Gaiaware AS
 * read the details at http://gaiaware.net/product/dual-licensing 
 ******************************************************************/

/*
 * **************************************************************
 * Author : Pavol Rusanov, Czech 
 * **************************************************************
*/

using System;
using System.Web;
using System.Text;
using System.Web.UI;
using System.Drawing;
using System.ComponentModel;
using System.Web.UI.HtmlControls;
using System.Collections.Generic;
using ASP = System.Web.UI.WebControls;

[assembly: WebResource("Gaia.WebWidgets.CommunityExtensions.Resources.blank.htm", "text/html")]
[assembly: WebResource("Gaia.WebWidgets.CommunityExtensions.Scripts.FileUpload.js", "text/javascript")]

namespace Gaia.WebWidgets.CommunityExtensions
{
    using HtmlFormatting;

    /// <summary>
    /// Ajax FileUpload control that allows you to upload many files with ajax technology. The Control
    /// was created by community member Pavol. 
    /// </summary>
    [ParseChildren(true)]
    [PersistChildren(false)]
    [ToolboxData("<{0}:FileUpload runat=\"server\" />")]
    [ToolboxBitmap(typeof(FileUpload), "Resources.Gaia.WebWidgets.CommunityExtensions.FileUpload.bmp")]
    public class FileUpload : Panel, IAjaxControl, IExtraPropertyCallbackRenderer
    {
        #region [ -- Private Members -- ]
        private bool _enabledChanged;
        private string _imgLoadingSrc = string.Empty;
        private bool _makeCallback;
        private int _maxFiles = 1;
        private string _textBtnUpload = "Save";
        private string _textDelete = "Delete";
        private string _textError = "Upload Failed";
        private string _textHeader = "Files to Upload";
        private string _textSizeError = "File Size Exceeded";
        private string _textUploaded = "Saved";
        private string _textUploading = "Uploading ...";
        private bool _uploadOnce;
        private readonly List<ASP.FileUpload> _listOfFileUploadControls = new List<ASP.FileUpload>();
        private string _blankPath = string.Empty;
        private readonly Button _callbackButton = new Button();
        private readonly HtmlGenericControl _uploadButton = new HtmlGenericControl("input");
        private readonly HiddenField _hiddenField = new HiddenField();
        private readonly HtmlGenericControl _iFrame = new HtmlGenericControl("iframe"); 
        #endregion

        #region [ -- Public Properties -- ]

        /// <summary>
        /// Display text while uploading files
        /// </summary>
        [Browsable(true)]
        [DefaultValue("Uploading ...")]
        public string TextUploading
        {
            get { return _textUploading; }
            set { _textUploading = value; }
        }

        /// <summary>
        /// Display text when file has been successfully uploaded
        /// </summary>
        [Browsable(true)]
        [DefaultValue("Saved")]
        public string TextUploaded
        {
            get { return _textUploaded; }
            set { _textUploaded = value; }
        }

        /// <summary>
        /// Display text when some error on the server occured.
        /// </summary>
        [Browsable(true)]
        [DefaultValue("Upload Failed")]
        public string TextError
        {
            get { return _textError; }
            set { _textError = value; }
        }

        /// <summary>
        /// Display text when file size has exceeded
        /// </summary>
        [Browsable(true)]
        [DefaultValue("File Size Exceeded")]
        public string TextSizeError
        {
            get { return _textSizeError; }
            set { _textSizeError = value; }
        }

        /// <summary>
        /// Display text for header
        /// </summary>
        [Browsable(true)]
        [DefaultValue("Files to Upload")]
        public string TextHeader
        {
            get { return _textHeader; }
            set { _textHeader = value; }
        }

        /// <summary>
        /// Save Button Text
        /// </summary>
        [Browsable(true)]
        [DefaultValue("Save")]
        public string TextBtnUpload
        {
            get { return _textBtnUpload; }
            set { _textBtnUpload = value; }
        }

        /// <summary>
        /// Delete Button Text
        /// </summary>
        [Browsable(true)]
        [DefaultValue("Delete")]
        public string TextDelete
        {
            get { return _textDelete; }
            set { _textDelete = value; }
        }

        /// <summary>
        /// How many files possible to upload at one moment
        /// </summary>
        [Browsable(true)]
        [DefaultValue(1)]
        public int MaxFiles
        {
            get { return _maxFiles; }
            set { _maxFiles = value; }
        }

        /// <summary>
        /// Set to true if you want the files only once and then the control becomes disabled
        /// </summary>
        [Browsable(true)]
        [DefaultValue(false)]
        public bool UploadOnce
        {
            get { return _uploadOnce; }
            set { _uploadOnce = value; }
        }

        /// <summary>
        /// Set to true if you want a callback after all files has been uploaded to the server.
        /// </summary>
        [Browsable(true)]
        [DefaultValue(false)]
        public bool MakeCallback
        {
            get { return _makeCallback; }
            set { _makeCallback = value; }
        }

        /// <summary>
        /// Source of the image loading.
        /// </summary>
        [Browsable(true)]
        [DefaultValue("")]
        public string ImgLoadingSrc
        {
            get { return _imgLoadingSrc; }
            set { _imgLoadingSrc = value; }
        }

        /// <summary>
        /// Enabled overridden property
        /// </summary>
        public override bool Enabled
        {
            get { return base.Enabled; }
            set
            {
                if (base.Enabled != value)
                    _enabledChanged = true;
                base.Enabled = value;
            }
        }

        #endregion

        #region [ -- FileUploadEventArgs -- ]

        /// <summary>
        /// EventArgs used for the FileUploaded Event. 
        /// </summary>
        public class FileUploadEventArgs : EventArgs
        {
            public FileUploadEventArgs(HttpPostedFile postedFile, string text)
            {
                PostedFile = postedFile;
                InfoText = text;
            }

            /// <summary>
            /// The Posted File
            /// </summary>
            public HttpPostedFile PostedFile { get; set; }

            /// <summary>
            /// Info Text
            /// </summary>
            public string InfoText { get; set; }
        }

        #endregion

        #region [ -- Event Handlers -- ]

        /// <summary>
        /// This Event is fired when the file file has been uploaded
        /// </summary>
        public event EventHandler<FileUploadEventArgs> FileUploaded;

        /// <summary>
        /// This Event is fired when all files are completely uploaded. You need to set the
        /// MakeCallback property to true for this event to fire. 
        /// </summary>
        public event EventHandler UploadFinished;

        #endregion

        #region [ -- Overriden base class methods --]

        protected override void OnInit(EventArgs e)
        {
            InitializeComponent();
            base.OnInit(e);
            EnsureChildControls();
        }

        private void InitializeComponent()
        {
            _iFrame.ID = "upFrame";
            Controls.Add(_iFrame);

            _iFrame.Attributes.Add("name", _iFrame.ClientID);
            _iFrame.Attributes.Add("frameborder", "0");
            _iFrame.Attributes.Add("scrolling", "no");
            _iFrame.Style["display"] = "none";

            for (var i = 0; i < _maxFiles; i++)
            {
                var fu = new ASP.FileUpload {ID = "f" + i};
                fu.Style["display"] = "none";
                _listOfFileUploadControls.Add(fu);
            }

            _uploadButton.ID = "btnUP";
            _uploadButton.Attributes.Add("type", "submit");
            _uploadButton.Attributes.Add("value", _textBtnUpload);

            _callbackButton.ID = "btnC";
            _callbackButton.Style["display"] = "none";
            _callbackButton.Click += BtnCallbackClick;

            _hiddenField.ID = "hf";
            _hiddenField.Value = "0";
        }

        protected override void CreateChildControls()
        {
            _iFrame.Controls.Add(new LiteralControl("&nbsp;"));

            foreach (ASP.FileUpload fileUpload in _listOfFileUploadControls)
                Controls.Add(fileUpload);

            Controls.Add(_uploadButton);
            Controls.Add(_callbackButton);
            Controls.Add(_hiddenField);

            base.CreateChildControls();
        }

        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);

            // very important attribute. The postback is targeted into blank.html window, which is embedded resource.
            _blankPath = Page.ClientScript.GetWebResourceUrl(typeof(FileUpload), "Gaia.WebWidgets.CommunityExtensions.Resources.blank.htm");
            _iFrame.Attributes.Add("src", _blankPath);

            CheckIfFileUploaded();
        }

        protected override void IncludeScriptFiles()
        {
            base.IncludeScriptFiles();
            Manager.Instance.AddInclusionOfFileFromResource("Gaia.WebWidgets.CommunityExtensions.Scripts.prototype.js", typeof (FileUpload), "");
            Manager.Instance.AddInclusionOfFileFromResource("Gaia.WebWidgets.CommunityExtensions.Scripts.FileUpload.js", typeof(FileUpload), "Gaia.Extensions.FileUpload.browserFinishedLoading");
        }

        protected override void RenderControlHtml(XhtmlTagFactory create)
        {
            using (Tag div = create.Div(ClientID, CssClass))
            {
                Css.SerializeAttributesAndStyles(this, div);
                _iFrame.RenderControl(create.GetHtmlTextWriter());

                for (int i = 0; i < _listOfFileUploadControls.Count; i++)
                    _listOfFileUploadControls[i].RenderControl(create.GetHtmlTextWriter());

                _uploadButton.RenderControl(create.GetHtmlTextWriter());
                _callbackButton.RenderControl(create.GetHtmlTextWriter());
                _hiddenField.RenderControl(create.GetHtmlTextWriter());
                using (create.Div(ClientID + "inner", CombineCssClass("fileupload-inner")))
                {
                    using (Tag span = create.Span(ClientID + "inner-header", CombineCssClass("fileupload-inner-header")))
                    {
                        span.WriteContent(_textHeader);
                    }
                }
            }
        }

        #endregion

        #region [ -- Helper Functions -- ]

        private void CheckIfFileUploaded()
        {
            string hfValue = Page.Request.Form[_hiddenField.UniqueID];
            if (string.IsNullOrEmpty(hfValue) || hfValue != "1") return;
            string text = _textError;
            for (int i = 0; i < Page.Request.Files.Count; i++)
            {
                if (string.IsNullOrEmpty(Page.Request.Files[i].FileName) || Page.Request.Files[i].ContentLength <= 0)
                    continue;

                var args = new FileUploadEventArgs(Page.Request.Files[i], null);
                if (FileUploaded != null)
                    FileUploaded(this, args);

                text = string.IsNullOrEmpty(args.InfoText) ? _textUploaded : args.InfoText;
                break;
            }

            Page.Title = "upload|" + text;
            Page.Form.Visible = false;
        }

        /// <summary>
        /// After all files has been uploaded you can catch UploadFinished event
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        protected void BtnCallbackClick(object sender, EventArgs e)
        {
            if (UploadFinished != null)
                UploadFinished(this, EventArgs.Empty);
        }

        private string CombineCssClass(params string[] cssclass)
        {
            return Css.Combine(CssClass, cssclass);
        }

        #endregion

        #region [ -- Overridden IAjaxControl Methods -- ]

        PropertyStateManagerControl IAjaxControl.CreateControlStateManager()
        {
            return new PropertyStateManagerWebControl(this, ClientID, this);
        }

        string IAjaxControl.GetScript()
        {
            return new RegisterControl("Gaia.Extensions.FileUpload", ClientID)
                .AddProperty("enabled", Enabled)
                .AddProperty("frameID", _iFrame.ClientID)
                .AddProperty("btnID", _uploadButton.ClientID)
                .AddProperty("hfID", _hiddenField.ClientID)
                .AddProperty("tUploading", _textUploading)
                .AddProperty("tSizeError", _textSizeError)
                .AddProperty("tDelete", _textDelete)
                .AddProperty("maxFiles", _maxFiles)
                .AddProperty("uploadOnce", _uploadOnce)
                .AddProperty("divFilesID", ClientID + "inner")
                .AddProperty("imgSrc", _imgLoadingSrc)
                .AddProperty("btnCallbackID", _callbackButton.ClientID)
                .AddProperty("makeCallback", _makeCallback)
                .AddPropertyIfTrue(!string.IsNullOrEmpty(_blankPath), "blankPath", _blankPath)
                .ToString();
        }

        #endregion

        #region IExtraPropertyCallbackRenderer Members

        public void InjectPropertyChangesToCallbackResponse(StringBuilder code)
        {
            if (_enabledChanged)
            {
                code.Append(".onEnabledChanged(" + Enabled.ToString().ToLower() + ")");
            }
        }

        #endregion

    }
}

File: Extras/Gaia.Ajax.GPL/4.0 - 31.10.2013/Source/Gaia.WebWidgets.CommunityExtensions/Scripts/FileUpload.js
// Gaia Ajax Copyright (C) 2008 - 2011 Gaiaware AS. details at http://gaiaware.net/

/* 
 * Gaia Ajax - Ajax Control Library for ASP.NET
 * Copyright (C) 2008 - 2011 Gaiaware AS
 * All rights reserved.
 * This program is distributed under either GPL version 3 
 * as published by the Free Software Foundation or the
 * Gaia Commercial License version 1 as published by
 * Gaiaware AS
 * read the details at http://gaiaware.net
 */

/* ---------------------------------------------------------------------------
   Ajax FileUpload Control 
   Author : Pavol Rusanov, Czech 
   --------------------------------------------------------------------------- */

if (!Gaia.Extensions)
    Gaia.Extensions = Class.create();

Gaia.Extensions.FileUpload = jsface.Class(Gaia.Panel, {
    constructor: function(element, options) {
        Gaia.Extensions.FileUpload.$super.call(this, element, options);
        this.initializeFileUpload(element, options);
    },

    initializeFileUpload: function(element, options) {

        // disable upload button a catch click event
        var btn = $(this.options.btnID);
        btn.FUP = this;
        btn.disabled = true;
        Event.observe(btn, "click", this.uploadClick);


        this.element[0].setAttribute('name', 'fup');
        this.element[0].FUP = this;
        this.options.actualFU = -1;
        this.options.fileRecords = new Array();
        this.options.filesUploaded = false;

        this.createFiles();

        var emptyFile = this.getEmptyObjFile();
        if (emptyFile != null) {
            this.options.actualFile = this.createInputFile(emptyFile.id, emptyFile.name);
            this.options.actualFile.disabled = !this.options.enabled;
        }

        $(this.options.btnCallbackID).disabled = false;

    },

    // create array of file ids
    createFiles: function() {
        this.options.files = new Array();

        for (var i = 0; i < this.options.maxFiles; i++) {
            var file = $(this.element[0].id + "_f" + i);
            var objFile = new Object();
            objFile.id = file.id;
            objFile.name = file.name;
            objFile.empty = true;
            this.options.files[i] = objFile;
            file.parentNode.removeChild(file);
        }
    },
    // create an empty INPUT element (file)
    createInputFile: function(id, name) {
        var file = document.createElement('input');
        file.type = 'file';
        file.id = id;
        file.name = name;
        this.element[0].insertBefore(file, $(this.options.btnID));
        file.FUP = this;
        this.options.actualFile = file;
        Event.observe(file, "change", this.fileChanged);
        return file;
    },
    // the file has been chosen, create some info about in DIV below
    createFileRecord: function(file) {

        var css = this.element[0].className;

        var container = $(this.options.divFilesID);

        var div = document.createElement('div');
        div.file = file;

        var textSpan = document.createElement('span');
        textSpan.className = css + '-fileupload-file-text';
        textSpan.innerHTML = file.value;
        div.appendChild(textSpan);

        var btnDel = document.createElement('input');
        btnDel.type = 'submit';
        btnDel.value = this.options.tDelete;
        btnDel.className = css + '-fileupload-button-delete';
        btnDel.div = div;
        btnDel.FUP = this;
        Event.observe(btnDel, "click", this.btnDelClick);
        div.appendChild(btnDel);
        div.btnDel = btnDel;

        if (this.options.imgSrc != "") {
            var imgLoading = document.createElement('img');
            imgLoading.style.display = 'none';
            imgLoading.setAttribute("alt", "loading...");
            imgLoading.setAttribute("src", this.options.imgSrc);
            imgLoading.className = css + '-fileupload-img-loading';
            div.appendChild(imgLoading);
            div.imgSrc = imgLoading;
        }

        var infoSpan = document.createElement('span');
        infoSpan.className = css + '-fileupload-info-text';
        div.appendChild(infoSpan);
        div.infoSpan = infoSpan;

        container.appendChild(div);
        return div;
    },

    // the file has been chosen
    fileChanged: function() {

        $(this.FUP.options.btnID).disabled = false;
        this.style.display = 'none';

        if (this.FUP.options.filesUploaded) {
            this.FUP.deleteFileRecords();
            this.FUP.options.filesUploaded = false;
        }

        var div = this.FUP.createFileRecord(this);

        this.FUP.setObjFileEmpty(this.id, false);
        this.FUP.options.fileRecords[this.FUP.options.fileRecords.length] = div;

        var emptyFile = this.FUP.getEmptyObjFile();
        if (emptyFile != null) {
            this.FUP.options.actualFile = this.FUP.createInputFile(emptyFile.id, emptyFile.name);
        } else {
            this.FUP.options.actualFile = this.FUP.createInputFile(this.FUP.element.id + "ff", this.FUP.element.id + "ff");
            this.FUP.options.actualFile.disabled = true;
        }
    },
    // delete button has been clicked
    btnDelClick: function() {
        var container = $(this.FUP.options.divFilesID);
        container.removeChild(this.div);

        this.FUP.options.fileRecords.remove(this.div);
        this.FUP.setObjFileEmpty(this.div.file.id, true);
        this.div.file.parentNode.removeChild(this.FUP.options.actualFile);
        this.div.file.parentNode.removeChild(this.div.file);

        var emptyFile = this.FUP.getEmptyObjFile();
        if (emptyFile != null) {
            this.FUP.createInputFile(emptyFile.id, emptyFile.name);
        }
    },
    // upload button has been clicked or another file has to be uploaded
    uploadClick: function(infoText) {

        var fup = this.FUP == null ? this : this.FUP;

        var isMaxReached = fup.hideFileUploads(infoText);
        // is any file waiting in the list that has to be uploaded?
        if (!isMaxReached) {

            fup.prepareAndSubmitForm();

            var executer = new PeriodicalExecuter(fup.isUploadCompleted, 1);
            executer.FUP = fup;
        } else {
            fup.options.actualFU = -1;
            fup.deleteFileInputs(true);
            fup.setAllObjFileEmpty(true);
            fup.recoverOtherElementsFileUploads();
            fup.options.filesUploaded = true;
            fup.options.fileRecords.length = 0;
            var emptyFile = fup.getEmptyObjFile();
            if (emptyFile != null)
                fup.createInputFile(emptyFile.id, emptyFile.name);

            if (fup.options.uploadOnce)
                fup.options.actualFile.disabled = true;

            if (fup.options.makeCallback)
                $(fup.options.btnCallbackID).click();
        }
    },

    // some attributes that has to be filled
    prepareAndSubmitForm: function() {
        document.forms[0].target = this.options.frameID;
        document.forms[0].enctype = 'multipart/form-data';
        document.forms[0].encoding = 'multipart/form-data'; // IE hack
        //$(this.options.lblInfoID).innerHTML = this.options.tUploading;
        $(this.options.hfID).value = '1';

        this.options.lastTitle = 'blank';

        $(this.options.btnID).disabled = true;
        document.forms[0].submit();
    },
    // timer has ticked
    isUploadCompleted: function() {
        try {
            var frame = $(this.FUP.options.frameID);
            var newTitle = frame.contentWindow.document.title;
            // no message from server yet
            if (this.FUP.options.lastTitle == newTitle) {
                // contine timing...
            }
                // the file has been successfully uploaded
            else if (newTitle.indexOf('upload') != -1) {
                this.stop();
                document.forms[0].target = '';
                $(this.FUP.options.hfID).value = '0';
                frame.contentWindow.document.title = 'blank';

                var titles = newTitle.split('|');
                if (titles.length == 2)
                    this.FUP.uploadClick(titles[1]);
            } else throw ('error');
        }
            // the size of the file has been probably exceeded
        catch(err) {
            this.stop();
            frame.src = this.FUP.options.blankPath;
            document.forms[0].target = '';
            $(this.FUP.options.hfID).value = '0';
            this.FUP.uploadClick(this.FUP.options.tSizeError);
        }
    },
    // server wants to Enable/Disable the control
    onEnabledChanged: function(enabled) {
        this.options.actualFile.disabled = !enabled;
        return this;
    },
    // from here it's the helper functions like hiding, recovering file controls...
    setObjFileEmpty: function(id, empty) {
        for (var i = 0; i < this.options.files.length; i++) {
            if (this.options.files[i].id == id) {
                this.options.files[i].empty = empty;
                break;
            }
        }
    },

    setAllObjFileEmpty: function(empty) {
        for (var i = 0; i < this.options.files.length; i++) {
            this.options.files[i].empty = empty;
        }
    },

    getEmptyObjFile: function() {
        for (var i = 0; i < this.options.files.length; i++) {
            if (this.options.files[i].empty)
                return this.options.files[i];
        }
        return null;
    },

    deleteFileInputs: function(deleteActual) {
        var btn = $(this.options.btnID);
        for (var i = 0; i < this.options.fileRecords.length; i++) {
            var file = this.options.fileRecords[i].file;
            if (file.parentNode != null)
                file.parentNode.removeChild(file);
        }
        if (deleteActual && this.options.actualFile.parentNode != null)
            this.options.actualFile.parentNode.removeChild(this.options.actualFile);
    },

    recoverFileInputs: function() {
        var btn = $(this.options.btnID);
        for (var i = 0; i < this.options.fileRecords.length; i++) {
            var file = this.options.fileRecords[i].file;
            if (file.parentNode == null)
                this.element[0].insertBefore(file, btn);
        }
    },

    deleteFileRecords: function() {
        var divFiles = document.getElementById(this.options.divFilesID);
        if (divFiles.hasChildNodes()) {
            while (divFiles.childNodes.length >= 1 && divFiles.lastChild.tagName.toLowerCase() == 'div') {
                divFiles.removeChild(divFiles.lastChild);
            }
        }
        this.options.fileRecords.length = 0;
    },

    hideFileUploads: function(infoText) {

        if (this.options.actualFU > -1) {
            this.options.fileRecords[this.options.actualFU].infoSpan.innerHTML = infoText;
            if (this.options.imgSrc != "")
                this.options.fileRecords[this.options.actualFU].imgSrc.style.display = 'none';
        } else {
            this.hideOtherElementsFileUploads();
        }

        this.options.actualFU++;
        this.deleteFileInputs(false);

        if (this.options.fileRecords.length > this.options.actualFU) {
            var div = this.options.fileRecords[this.options.actualFU];

            this.element[0].insertBefore(div.file, $(this.options.btnID));
            div.infoSpan.innerHTML = this.options.tUploading;
            div.btnDel.disabled = true;
            if (this.options.imgSrc != "")
                div.imgSrc.style.display = '';
            return false;
        }
        return true;
    },

    hideOtherElementsFileUploads: function() {

        var FUPs = this.getElementsByName_iefix('div', 'fup');
        for (var i = 0; i < FUPs.length; i++) {
            if (FUPs[i].id != this.element[0].id)
                FUPs[i].FUP.deleteFileInputs(false);
        }
    },

    recoverOtherElementsFileUploads: function() {

        var FUPs = this.getElementsByName_iefix('div', 'fup');
        for (var i = 0; i < FUPs.length; i++) {
            if (FUPs[i].id != this.element[0].id)
                FUPs[i].FUP.recoverFileInputs();
        }
    },

    getElementsByName_iefix: function(tag, name) {

        var elem = document.getElementsByTagName(tag);
        var arr = new Array();
        for (i = 0, iarr = 0; i < elem.length; i++) {
            att = elem[i].getAttribute("name");
            if (att == name) {
                arr[iarr] = elem[i];
                iarr++;
            }
        }
        return arr;
    }
});

Array.prototype.remove = function(s) {
    for (i = 0; i < this.length; i++) {
        if (s == this[i]) this.splice(i, 1);
    }
}

